`exportc` mangles as C++ and doesn't work with `importc`(C) with `nim cpp`

Nim should allow controlling symbol mangling; it currently seems impossible with `nim cpp`.

Using test example below:
```
# this works
nim c -r main.nim

# this errors: Undefined symbols for architecture x86_64: _my_foobar1
nim cpp -r main.nim
```

## test example
main.nim:
```nim
proc foobar1(a: int) {.importc: "my_foobar1".}
proc main() = foobar1(10)
main()

# intentially putting this at the end;
# other use cases: this could be linked from a shared library (nim plugin)
# other use cases: forward declaration to avoid module import cycles
import ./t0202b
```
t0202b.nim:
```
proc foobar1(a: int) {.exportc: "my_foobar1".} = echo (a:a)
```


## very partial workaround
Here's my current workaround that works with `nim c` and `nim cpp`
It still doesn't let me mangle a c++ symbol the way I want, so it doesn't help if I need to export a symbol for a non-nim program that expects a given mangling over which I have no control. 
```
# main.nim:
proc getMangle(cName, cppName: string): string =
  when defined(cpp): cppName else: cName

proc foobar1(a: int) {.importc: getMangle("my_foobar1", "_Z10my_foobar1x").}
...

# t0202b.nim:
proc foobar1(a: int) {.exportc: "my_foobar1".} = echo (a:a)
```


## workaround suggested by @araq: I couldn't make it work
> you need to play tricks with emit "extern C" or .codegenDecl

no success, please explain

* I tried:
```
var foobar2 = proc(a: int) {.codegenDecl: "$1 foobar2_b $3".} = discard

# gives compilation error for t0202b.cpp:
error: use of undeclared identifier 'colonanonymous__t4FhpLqaEhvD69cnCq6B8Sg'
        foobar2_tPSgPPlTA53EAR1UTPkutw = colonanonymous__t4FhpLqaEhvD69cnCq6B8Sg;
```


* I also tried:
```
proc foobar2(a: int) {.codegenDecl: "$1 my_foobar1 $3".} = echo (a:a)
# this elides `foobar2` from `t0202b.cpp` as the symbol is unused, so gives a link error for main

# but if i add this to avoid eliding the symbol:
if false: foobar2(0)
# then I get:
error: use of undeclared identifier 'foobar2_t4FhpLqaEhvD69cnCq6B8Sg'
```


## workaround via emit: couldn't make it work
```nim
{.emit:"""extern "C"""".}
proc foobar2(a: int) {.exportc: "nim_foobar2".} = echo (a:a)
```
when you compile, it seems to work (no error!) but doesn't actually work:
the generated cpp code is:
```
extern "C"
N_NIMCALL(void, nim_foobar2)(NI a);
....

// NOTE: no `extern "C"` here so it doesn't relate to above declaration and won't be visible by outside
N_NIMCALL(void, nim_foobar2)(NI a) {
       ...
	echoBinSafe(T1_, 1);
	popFrame();
}
```

## [EDIT] (credits: @araq)
this seems to work:
```
proc foobar1(a: int) {.exportc: "my_foobar1", codegenDecl: """extern "C" $1 $2 $3""".} = echo "ok1"
```
however, it breaks `nim c` compatibity; so, see proposal below.

## proposal

### rule 1
when `exportc` is specified, it should imply `extern "C"` so that it works seamlessly in tandem with `importc` ; that's the most common use case and the one that'll lead to least bugs when we need to support both  `nim c` and `nim cpp`  (either in isolation, or even when a `nim c` main links against a `nim cpp` shared library)
Eg:
```nim
# this should mangle to same symbol `my_foobar1` regardless we use `nim c`  or `nim cpp`
proc foobar1(a: int) {.exportc: "my_foobar1".} = echo (a:a)
# ditto, mangles to "foobar1"
proc foobar1(a: int) {.exportc.} = echo (a:a)
# ditto, mangles to "foofoobar1bar"
proc foobar1(a: int) {.exportc: "foo$1bar".} = echo (a:a)
```

### rule 2
We also allow (but as an opt-in option) for `exportc` to mangle as c++ as follows:
```nim
# this should mangle to same symbol `__Z10my_foobar1x` (`c++filt __Z10my_foobar1x  = my_foobar1(long long)`)
proc foobar1(a: int) {.exportc: "my_foobar1", mangling: cpp.} = echo (a:a)

# ditto with "$1" which only affects the function name before mangling via function arguments
```

## note:

* `{.push: mangling: cpp.}` can be used for convenience to wrap lots of procs at once
* in D, there is `pragma(mangle)` for this purpose;
it allows flexibilty, for example by conrolling how a particular type `Foo` gets mangled (say, to `FooAlias`), so that when a type `Bar` that depends on it (eg `std::vector<Foo>`) gets mangled (using c++ mangling rules), it'll use the name `FooAlias` inside the mangling of `Bar` , as if `Foo` was named `FooAlias` in the 1st place.
* [EDIT] `mangling` can support different options, eg:
  * `mangling:native` : use C or C++ mangling depending on whether `nim c` or `nim cpp` is used
  * `mangling: cpp`: use C++ mangling with `nim cpp`, and currently errors `with `nim c` but could add support in future (eg [link](https://stackoverflow.com/questions/40740604/how-do-i-get-the-mangled-name-of-a-nameddecl-in-clang) or by calling c++ codegen for the declaration along with its minimal needed dependencies )
  * `mangling: objc`, `mangling: d` (reserved for future use)

--- a/changelog.md
+++ b/changelog.md
@@ -31,6 +31,7 @@ type
 
 ```
 
+- `exportc` now uses C instead of C++ mangling with `nim cpp`, matching behavior of `importc`, see #10578
 
 ### Breaking changes in the compiler
 

--- a/changelog.md
+++ b/changelog.md
@@ -32,6 +32,7 @@ type
 ```
 
 - `exportc` now uses C instead of C++ mangling with `nim cpp`, matching behavior of `importc`, see #10578
+  Use the new `exportcpp` to mangle as C++ when using `nim cpp`.
 
 ### Breaking changes in the compiler
 

--- a/compiler/ast.nim
+++ b/compiler/ast.nim
@@ -239,6 +239,7 @@ type
     sfForward,        # symbol is forward declared
     sfImportc,        # symbol is external; imported
     sfExportc,        # symbol is exported (under a specified name)
+    sfMangleCpp,      # mangle as cpp (combines with `sfExportc`)
     sfVolatile,       # variable is volatile
     sfRegister,       # variable should be placed in a register
     sfPure,           # object is "pure" that means it has no type-information

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -1039,7 +1039,8 @@ proc requiresExternC(m: BModule; sym: PSym): bool {.inline.} =
   result = (sfCompileToCpp in m.module.flags and
            sfCompileToCpp notin sym.getModule().flags and
            m.config.cmd != cmdCompileToCpp) or (
-           sym.flags * {sfImportc, sfInfixCall, sfCompilerProc} == {sfImportc} and
+           sym.flags * {sfInfixCall, sfCompilerProc} == {} and
+           sym.flags * {sfImportc, sfExportc} != {} and
            sym.magic == mNone and
            m.config.cmd == cmdCompileToCpp)
 

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -1039,7 +1039,7 @@ proc requiresExternC(m: BModule; sym: PSym): bool {.inline.} =
   result = (sfCompileToCpp in m.module.flags and
            sfCompileToCpp notin sym.getModule().flags and
            m.config.cmd != cmdCompileToCpp) or (
-           sym.flags * {sfInfixCall, sfCompilerProc} == {} and
+           sym.flags * {sfInfixCall, sfCompilerProc, sfMangleCpp} == {} and
            sym.flags * {sfImportc, sfExportc} != {} and
            sym.magic == mNone and
            m.config.cmd == cmdCompileToCpp)

--- a/compiler/pragmas.nim
+++ b/compiler/pragmas.nim
@@ -19,25 +19,27 @@ const
   LastCallConv* = wNoconv
 
 const
-  procPragmas* = {FirstCallConv..LastCallConv, wImportc, wExportc, wNodecl,
+  declPragmas = {wImportc, wImportObjC, wImportCpp, wExportc, wExportCpp,
+    wExportNims, wExtern, wDeprecated, wNodecl, wError, wUsed}
+    ## common pragmas for declarations, to a good approximation
+  procPragmas* = declPragmas + {FirstCallConv..LastCallConv,
     wMagic, wNoSideEffect, wSideEffect, wNoreturn, wDynlib, wHeader,
-    wCompilerProc, wNonReloadable, wCore, wProcVar, wDeprecated, wVarargs, wCompileTime, wMerge,
-    wBorrow, wExtern, wImportCompilerProc, wThread, wImportCpp, wImportObjC,
-    wAsmNoStackFrame, wError, wDiscardable, wNoInit, wCodegenDecl,
+    wCompilerProc, wNonReloadable, wCore, wProcVar, wVarargs, wCompileTime, wMerge,
+    wBorrow, wImportCompilerProc, wThread,
+    wAsmNoStackFrame, wDiscardable, wNoInit, wCodegenDecl,
     wGensym, wInject, wRaises, wTags, wLocks, wDelegator, wGcSafe,
-    wConstructor, wExportNims, wUsed, wLiftLocals, wStackTrace, wLineTrace, wNoDestroy}
+    wConstructor, wLiftLocals, wStackTrace, wLineTrace, wNoDestroy}
   converterPragmas* = procPragmas - {wNoDestroy}
   methodPragmas* = procPragmas+{wBase}-{wImportCpp, wNoDestroy}
   templatePragmas* = {wDeprecated, wError, wGensym, wInject, wDirty,
     wDelegator, wExportNims, wUsed, wPragma}
-  macroPragmas* = {FirstCallConv..LastCallConv, wImportc, wExportc,
-    wNodecl, wMagic, wNoSideEffect, wCompilerProc, wNonReloadable, wCore, wDeprecated, wExtern,
-    wImportCpp, wImportObjC, wError, wDiscardable, wGensym, wInject, wDelegator,
-    wExportNims, wUsed}
-  iteratorPragmas* = {FirstCallConv..LastCallConv, wNoSideEffect, wSideEffect,
-    wImportc, wExportc, wNodecl, wMagic, wDeprecated, wBorrow, wExtern,
-    wImportCpp, wImportObjC, wError, wDiscardable, wGensym, wInject, wRaises,
-    wTags, wLocks, wGcSafe, wExportNims, wUsed}
+  macroPragmas* = declPragmas + {FirstCallConv..LastCallConv,
+    wMagic, wNoSideEffect, wCompilerProc, wNonReloadable, wCore,
+    wDiscardable, wGensym, wInject, wDelegator}
+  iteratorPragmas* = declPragmas + {FirstCallConv..LastCallConv, wNoSideEffect, wSideEffect,
+    wMagic, wBorrow,
+    wDiscardable, wGensym, wInject, wRaises,
+    wTags, wLocks, wGcSafe}
   exprPragmas* = {wLine, wLocks, wNoRewrite, wGcSafe, wNoSideEffect}
   stmtPragmas* = {wChecks, wObjChecks, wFieldChecks, wRangeChecks,
     wBoundChecks, wOverflowChecks, wNilChecks, wStyleChecks, wAssertions,
@@ -49,25 +51,25 @@ const
     wDeprecated,
     wFloatChecks, wInfChecks, wNanChecks, wPragma, wEmit, wUnroll,
     wLinearScanEnd, wPatterns, wTrMacros, wEffects, wNoForward, wReorder, wComputedGoto,
-    wInjectStmt, wDeprecated, wExperimental, wThis, wUsed}
-  lambdaPragmas* = {FirstCallConv..LastCallConv, wImportc, wExportc, wNodecl,
+    wInjectStmt, wExperimental, wThis, wUsed}
+  lambdaPragmas* = declPragmas + {FirstCallConv..LastCallConv,
     wNoSideEffect, wSideEffect, wNoreturn, wDynlib, wHeader,
-    wDeprecated, wExtern, wThread, wImportCpp, wImportObjC, wAsmNoStackFrame,
-    wRaises, wLocks, wTags, wGcSafe, wCodegenDecl}
-  typePragmas* = {wImportc, wExportc, wDeprecated, wMagic, wAcyclic, wNodecl,
-    wPure, wHeader, wCompilerProc, wCore, wFinal, wSize, wExtern, wShallow,
-    wImportCpp, wImportObjC, wError, wIncompleteStruct, wByCopy, wByRef,
+    wThread, wAsmNoStackFrame,
+    wRaises, wLocks, wTags, wGcSafe, wCodegenDecl} - {wExportNims, wError, wUsed}  # why exclude these?
+  typePragmas* = declPragmas + {wMagic, wAcyclic,
+    wPure, wHeader, wCompilerProc, wCore, wFinal, wSize, wShallow,
+    wIncompleteStruct, wByCopy, wByRef,
     wInheritable, wGensym, wInject, wRequiresInit, wUnchecked, wUnion, wPacked,
-    wBorrow, wGcSafe, wExportNims, wPartial, wUsed, wExplain, wPackage}
-  fieldPragmas* = {wImportc, wExportc, wDeprecated, wExtern,
-    wImportCpp, wImportObjC, wError, wGuard, wBitsize, wUsed}
-  varPragmas* = {wImportc, wExportc, wVolatile, wRegister, wThreadVar, wNodecl,
-    wMagic, wHeader, wDeprecated, wCompilerProc, wCore, wDynlib, wExtern,
-    wImportCpp, wImportObjC, wError, wNoInit, wCompileTime, wGlobal,
-    wGensym, wInject, wCodegenDecl, wGuard, wGoto, wExportNims, wUsed, wCursor}
-  constPragmas* = {wImportc, wExportc, wHeader, wDeprecated, wMagic, wNodecl,
-    wExtern, wImportCpp, wImportObjC, wError, wGensym, wInject, wExportNims,
-    wIntDefine, wStrDefine, wBoolDefine, wUsed, wCompilerProc, wCore}
+    wBorrow, wGcSafe, wPartial, wExplain, wPackage}
+  fieldPragmas* = declPragmas + {
+    wGuard, wBitsize} - {wExportNims, wNodecl} # why exclude these?
+  varPragmas* = declPragmas + {wVolatile, wRegister, wThreadVar,
+    wMagic, wHeader, wCompilerProc, wCore, wDynlib,
+    wNoInit, wCompileTime, wGlobal,
+    wGensym, wInject, wCodegenDecl, wGuard, wGoto, wCursor}
+  constPragmas* = declPragmas + {wHeader, wMagic,
+    wGensym, wInject,
+    wIntDefine, wStrDefine, wBoolDefine, wCompilerProc, wCore}
   letPragmas* = varPragmas
   procTypePragmas* = {FirstCallConv..LastCallConv, wVarargs, wNoSideEffect,
                       wThread, wRaises, wLocks, wTags, wGcSafe}
@@ -775,8 +777,13 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: var int,
       if {optStyleHint, optStyleError} * c.config.globalOptions != {}:
         checkPragmaUse(c.config, key.info, k, ident.s)
       case k
-      of wExportc:
+      of wExportc, wExportCpp:
         makeExternExport(c, sym, getOptionalStr(c, it, "$1"), it.info)
+        if k == wExportCpp:
+          if c.config.cmd != cmdCompileToCpp:
+            localError(c.config, it.info, "exportcpp requires `nim cpp`, got " & $c.config.cmd)
+          else:
+            incl(sym.flags, sfMangleCpp)
         incl(sym.flags, sfUsed) # avoid wrong hints
       of wImportc:
         let name = getOptionalStr(c, it, "$1")

--- a/compiler/wordrecg.nim
+++ b/compiler/wordrecg.nim
@@ -42,7 +42,7 @@ type
     wImmediate, wConstructor, wDestructor, wDelegator, wOverride,
     wImportCpp, wImportObjC,
     wImportCompilerProc,
-    wImportc, wExportc, wExportNims, wIncompleteStruct, wRequiresInit,
+    wImportc, wExportc, wExportCpp, wExportNims, wIncompleteStruct, wRequiresInit,
     wAlign, wNodecl, wPure, wSideEffect, wHeader,
     wNoSideEffect, wGcSafe, wNoreturn, wMerge, wLib, wDynlib,
     wCompilerProc, wCore, wProcVar, wBase, wUsed,
@@ -129,7 +129,7 @@ const
 
     "immediate", "constructor", "destructor", "delegator", "override",
     "importcpp", "importobjc",
-    "importcompilerproc", "importc", "exportc", "exportnims",
+    "importcompilerproc", "importc", "exportc", "exportcpp", "exportnims",
     "incompletestruct",
     "requiresinit", "align", "nodecl", "pure", "sideeffect",
     "header", "nosideeffect", "gcsafe", "noreturn", "merge", "lib", "dynlib",

--- a/tests/cpp/mexportc.nim
+++ b/tests/cpp/mexportc.nim
@@ -4,3 +4,6 @@ var fun0 {.exportc.} = 10
 proc fun1() {.exportc.} = discard
 proc fun2() {.exportc: "$1".} = discard
 proc fun3() {.exportc: "fun3Bis".} = discard
+
+when defined cpp:
+  proc funx1() {.exportcpp.} = discard

--- a/tests/cpp/mexportc.nim
+++ b/tests/cpp/mexportc.nim
@@ -1,4 +1,6 @@
 {.used.} # ideally, would not be needed
 
+var fun0 {.exportc.} = 10
 proc fun1() {.exportc.} = discard
-var fun2 {.exportc.} = 10
+proc fun2() {.exportc: "$1".} = discard
+proc fun3() {.exportc: "fun3Bis".} = discard

--- a/tests/cpp/mexportc.nim
+++ b/tests/cpp/mexportc.nim
@@ -1,0 +1,4 @@
+{.used.} # ideally, would not be needed
+
+proc fun1() {.exportc.} = discard
+var fun2 {.exportc.} = 10

--- a/tests/cpp/texportc.nim
+++ b/tests/cpp/texportc.nim
@@ -7,9 +7,16 @@ proc fun1() {.importc.}
 proc fun2() {.importc: "$1".}
 proc fun3() {.importc: "fun3Bis".}
 
+when defined cpp:
+  # proc funx1() {.importcpp.} # this does not work yet
+  proc funx1() {.importc: "_Z5funx1v".}
+
 doAssert fun0 == 10
 fun1()
 fun2()
 fun3()
+
+when defined cpp:
+  funx1()
 
 import ./mexportc

--- a/tests/cpp/texportc.nim
+++ b/tests/cpp/texportc.nim
@@ -1,0 +1,11 @@
+discard """
+  targets: "c cpp"
+"""
+
+var fun2 {.importc.}: int
+proc fun1(){.importc.}
+
+fun1()
+doAssert fun2 == 10
+
+import ./mexportc

--- a/tests/cpp/texportc.nim
+++ b/tests/cpp/texportc.nim
@@ -2,10 +2,14 @@ discard """
   targets: "c cpp"
 """
 
-var fun2 {.importc.}: int
-proc fun1(){.importc.}
+var fun0 {.importc.}: int
+proc fun1() {.importc.}
+proc fun2() {.importc: "$1".}
+proc fun3() {.importc: "fun3Bis".}
 
+doAssert fun0 == 10
 fun1()
-doAssert fun2 == 10
+fun2()
+fun3()
 
 import ./mexportc

